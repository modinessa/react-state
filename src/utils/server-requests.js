//const url = 'http://localhost:3000';
const url = '';

export function getUsers() {
  return fetch(`${url}/community`, {
    method: 'GET',
  });
}

export function getUser(id) {
	return fetch(`/community/${id}`);
}

export function subscribe(email) {
  return fetch(`${url}/subscribe`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      email: `${email}`,
    }),
  });
}

export function unsubscribe() {
  return fetch(`${url}/unsubscribe`, { method: 'POST' });
}
