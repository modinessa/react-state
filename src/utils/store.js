import { configureStore } from "@reduxjs/toolkit";
import usersReducer from "../components/reducer";

export const store = configureStore({
	reducer: {
		users: usersReducer,
	}
});