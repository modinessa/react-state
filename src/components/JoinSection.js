import React from "react";
import { useForm } from "react-hook-form";
import { subscribe, unsubscribe } from "../utils/server-requests";
import * as constants from "../constants/constants";
import { yupResolver } from "@hookform/resolvers/yup";
import { createRegExp } from "../utils/email-validator";
import * as yup from "yup";
import { setIsSubscribed } from "./reducer";
import { useSelector, useDispatch } from "react-redux";

const schema = yup.object({
  email: yup.string().required("Please Enter your Email")
						.matches(createRegExp(constants.VALID_EMAIL_ENDINGS),
						'Your Email must end with "gmail.com", "outlook.com", or "yandex.ru"')});

export function JoinSection() {
	const dispatch = useDispatch();
	const {isSubscribed} = useSelector((state) => state.users);
	const {
		register,
		handleSubmit,
		formState
	} = useForm(
		{resolver: yupResolver(schema)}
	);
	const { errors, isSubmitting } = formState;

	const onSubmit = (data) => {
		if(!isSubscribed) {
			if (!errors.email) {
				subscribe(data.email)
				.then((response) => {
					if (!response.ok) {
						response.json()
							.then((error) => {
								window.alert(error.error);
							})
					} else {	
						dispatch(setIsSubscribed(!isSubscribed));
					}
				})
			} 	
		} else {	
			unsubscribe();
			dispatch(setIsSubscribed(!isSubscribed));
		}
	};

	const onError = (errors) => {
		window.alert(errors.email.message);
	}

	return (
		<section className="app-section app-section--image-joun-us">
			<h2 className="app-title">
      	Join Our Program
    	</h2>
			<h3 className="app-subtitle">
      	Sed do eiusmod tempor incididunt<br/>
				ut labore et dolore magna aliqua.
    	</h3>
			<form className={`app-section--form-join-us
											${isSubscribed ? "unsubscribe" : ""} `}
						name="EmailForm"
						onSubmit={handleSubmit(onSubmit, onError)}>
				<input {...register("email")} className="app-section--email-join-us"
								type="email" id="user-email" name="email" placeholder="Email"/>
				<button id="subBtn" type="submit"
								className="app-section__button app-section__join-us--button"
								disabled={isSubmitting}>
					{!isSubscribed ? "Subscribe" : "Unsubscribe"}
				</button>
			</form>
		</section>
	)
}
