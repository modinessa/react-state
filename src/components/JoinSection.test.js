import {Provider} from "react-redux";
import {render, fireEvent} from '@testing-library/react';
import renderer from 'react-test-renderer';
import {rest} from 'msw';
import {setupServer} from 'msw/node';

import {store} from "../utils/store";
import {JoinSection} from './JoinSection';

const VALID_EMAIL = 'valid@gmail.com'
const NOT_VALID_EMAIL = 'dont_care@gogol.ru';

const mockServer = setupServer(
	rest.post('/subscribe',
		(req, res, ctx) =>
		  res(ctx.status(req.body.email === VALID_EMAIL ? 200 : 401))),
	rest.post('/unsubscribe', (req, res, ctx) => res()),
)

describe('JoinSection', () => {
	let joinSection;
	let input;
	let submitButton;
	
	beforeAll(() => {
		mockServer.listen();
	});

	afterAll(() => {
		mockServer.close();
	});

	beforeEach(() => {
		joinSection = render(
			<Provider store={store}>
				<JoinSection/>
			</Provider>
		);
		input = joinSection.getByLabelText('user-email-input');
		submitButton = joinSection.getByLabelText('submit-button');
	});

	it('renders component', () => {
		expect(renderer.create(
			<Provider store={store}>
				<JoinSection/>
			</Provider>)).toMatchSnapshot();
	});

	it('submits valid email', async () => {
		let btn;
		expect(store.getState().users.isSubscribed).toEqual(false);

		fireEvent.change(input, {target: {value: VALID_EMAIL}});
		fireEvent.click(submitButton);

		btn = await joinSection.findByText('Unsubscribe');

		expect(btn).not.toBeUndefined();
		expect(store.getState().users.isSubscribed).toEqual(true);

		fireEvent.click(btn);

		btn = await joinSection.findByText('Subscribe');

		expect(btn).not.toBeUndefined();
		expect(store.getState().users.isSubscribed).toEqual(false);
	});

	it('submits invalid email', async () => {
		expect(store.getState().users.isSubscribed).toEqual(false);

		fireEvent.change(input, {target: {value: NOT_VALID_EMAIL}});
		fireEvent.click(submitButton);

		const btn = await joinSection.findByText('Subscribe');

		expect(btn).not.toBeUndefined();
		expect(store.getState().users.isSubscribed).toEqual(false);
	});
});
