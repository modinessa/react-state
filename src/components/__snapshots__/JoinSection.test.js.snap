// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinSection renders component 1`] = `
<section
  className="app-section app-section--image-joun-us"
>
  <h2
    className="app-title"
  >
    Join Our Program
  </h2>
  <h3
    className="app-subtitle"
  >
    Sed do eiusmod tempor incididunt
    <br />
    ut labore et dolore magna aliqua.
  </h3>
  <form
    className="app-section--form-join-us
											 "
    name="EmailForm"
    onSubmit={[Function]}
  >
    <input
      aria-label="user-email-input"
      className="app-section--email-join-us"
      id="user-email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Email"
      type="email"
    />
    <button
      aria-label="submit-button"
      className="app-section__button app-section__join-us--button"
      disabled={false}
      id="subBtn"
      type="submit"
    >
      Subscribe
    </button>
  </form>
</section>
`;
